File: packages/sdk-ts/src/stagehand.ts
import { RPCClient } from "./rpcClient.js";
import {
  DefaultExtractDataSchema,
  MAX_CALLBACK_BATCH_TIMEOUT_MS,
  STAGEHAND_PROTOCOL_VERSION,
  StagehandInitParamsSchema,
} from "@browserbasehq/stagehand-protocol/schemas";
import { JSONRPCErrorObjectSchema } from "@browserbasehq/stagehand-protocol/json-rpc/schemas";
import {
  StagehandMethods,
  StagehandNotifications,
} from "@browserbasehq/stagehand-protocol/schema-registry";
import type {
  Action,
  ActResult,
  CallbackBatchResult,
  DefaultExtractData,
  ObserveResult,
  StagehandMetrics,
  StagehandRpcNotification,
} from "@browserbasehq/stagehand-protocol/types";
import { z } from "zod/v4";
import { BrowserContext } from "./browserContext.js";
import { serializeClientLocatorOptions } from "./clientLocatorOptions.js";
import {
  StagehandClientActOptionsSchema,
  StagehandClientExtractOptionsSchema,
  StagehandCreateOptionsSchema,
  StagehandClientObserveOptionsSchema,
  type StagehandClientActOptions,
  type StagehandClientExtractOptions,
  type ResolvedStagehandClientLoggingConfig,
  type ResolvedStagehandClientCreateConfig,
  type StagehandCreateOptions,
  type StagehandClientObserveOptions,
} from "./clientSchemas.js";
import { CDPConnectionClosedError } from "./cdpClient.js";
import { STAGEHAND_SDK_CLIENT_INFO } from "./sdkIdentity.js";
import {
  claimStagehandBrowser,
  invalidateStagehandBrowser,
  releaseStagehandBrowser,
  type ClaimedStagehandBrowser,
  type StagehandBrowser,
} from "./browser/factories.js";
import { attachStagehandBrowserContext, detachStagehandBrowserContext } from "./browser/index.js";
import { withStagehandInitDeadline } from "./timeouts.js";
import type { ExperimentalBatchCallback, ExperimentalBatchOptions } from "./batch.js";

type ProtocolExtractResult = import("@browserbasehq/stagehand-protocol/types").ExtractResult;

export type ExtractResult<Schema extends z.ZodType> = Omit<ProtocolExtractResult, "data"> & {
  data: z.output<Schema>;
};

const isZodSchema = (value: unknown): value is z.ZodType =>
  typeof value === "object" &&
  value !== null &&
  "parse" in value &&
  typeof value.parse === "function" &&
  "safeParse" in value &&
  typeof value.safeParse === "function";

const nativeFunctionSourcePattern =
  /^\s*function(?:\s+[^()]*)?\([^)]*\)\s*\{\s*\[native code\]\s*\}\s*$/;

export class Stagehand {
  isInitialized = false;
  rpcClient: RPCClient | undefined;
  removeNotificationListener: (() => void) | undefined;
  removeClientLLMHandler: (() => void) | undefined;
  closePromise: Promise<void> | undefined;
  initRequestStarted = false;

  private constructor(
    private readonly browserHandle: StagehandBrowser,
    private readonly createConfig: ResolvedStagehandClientCreateConfig,
  ) {}

  static async create(input: StagehandCreateOptions): Promise<Stagehand> {
    const { browser, ...createConfig } = StagehandCreateOptionsSchema.parse(input);
    const claimedBrowser = claimStagehandBrowser(browser);
    const stagehand = new Stagehand(browser, createConfig);
    let lifecycleSignal: AbortSignal | undefined;
    try {
      await withStagehandInitDeadline((signal) => {
        lifecycleSignal = signal;
        return stagehand.initialize(claimedBrowser, signal);
      });
      return stagehand;
    } catch (error) {
      const initFailureIsAmbiguous =
        lifecycleSignal?.aborted ||
        (stagehand.initRequestStarted && !isDefinitiveRPCErrorResponse(error));
      if (initFailureIsAmbiguous) {
        try {
          await invalidateStagehandBrowser(browser);
        } catch (cleanupError) {
          throw new AggregateError(
            [error, cleanupError],
            "Stagehand initialization failed ambiguously and browser invalidation also failed",
            { cause: error },
          );
        }
      } else {
        releaseStagehandBrowser(browser);
      }
      throw error;
    }
  }

  get browser(): StagehandBrowser {
    return this.browserHandle;
  }

  get initialized(): boolean {
    return this.isInitialized;
  }

  async metrics(): Promise<StagehandMetrics> {
    return this.connectedRpcClient.send(StagehandMethods.stagehandMetrics, {});
  }

  async experimentalBatch<Result>(
    callback: ExperimentalBatchCallback<undefined, Result>,
  ): Promise<Awaited<Result>>;
  async experimentalBatch<Input, Result>(
    callback: ExperimentalBatchCallback<Input, Result>,
    input: Input,
    options?: ExperimentalBatchOptions,
  ): Promise<Awaited<Result>>;
  async experimentalBatch<Input, Result>(
    callback: ExperimentalBatchCallback<Input, Result>,
    input?: Input,
    options: ExperimentalBatchOptions = {},
  ): Promise<Awaited<Result>> {
    if (typeof callback !== "function") {
      throw new TypeError("stagehand.experimentalBatch() requires a callback function");
    }
    if (options === null || typeof options !== "object" || Array.isArray(options)) {
      throw new TypeError("stagehand.experimentalBatch() options must be an object");
    }
    const parsedInput = input === undefined ? undefined : z.json().parse(input);
    const timeout = options.timeout ?? 30_000;
    if (!Number.isInteger(timeout) || timeout <= 0) {
      throw new RangeError("stagehand.experimentalBatch() timeout must be a positive integer");
    }
    if (timeout > MAX_CALLBACK_BATCH_TIMEOUT_MS) {
      throw new RangeError(
        `stagehand.experimentalBatch() timeout must not exceed ${MAX_CALLBACK_BATCH_TIMEOUT_MS} milliseconds`,
      );
    }
    const callbackSource = Function.prototype.toString.call(callback);
    if (nativeFunctionSourcePattern.test(callbackSource)) {
      throw new TypeError("stagehand.experimentalBatch() callback must be serializable JavaScript");
    }

    const result: CallbackBatchResult = await this.connectedRpcClient.send(
      StagehandMethods.stagehandCallbackBatch,
      {
        callbackSource,
        ...(parsedInput === undefined ? {} : { input: parsedInput }),
        options: {
          ...(options.page ? { pageId: options.page.pageId } : {}),
          timeout,
        },
      },
    );
    return result.value as Awaited<Result>;
  }

  private async initialize(browser: ClaimedStagehandBrowser, signal: AbortSignal): Promise<void> {
    const createConfig = this.createConfig;
    const rpcClient = new RPCClient(browser.cdpClient);
    this.rpcClient = rpcClient;

    try {
      this.removeNotificationListener = rpcClient.onNotification((notification) =>
        handleStagehandNotification(notification, createConfig.logging),
      );
      if (createConfig.model && "generate" in createConfig.model) {
        this.removeClientLLMHandler = rpcClient.onRequest(
          StagehandMethods.llmGenerate,
          createConfig.model.generate,
        );
      }

      this.initRequestStarted = true;
      await rpcClient.sendStagehandInit(
        stagehandCreateParamsForWorker(createConfig, browser),
        signal,
      );
      attachStagehandBrowserContext(
        this.browserHandle,
        new BrowserContext(rpcClient, () => this.browserHandle.close()),
      );
    } catch (error) {
      this.removeClientLLMHandler?.();
      this.removeClientLLMHandler = undefined;
      this.removeNotificationListener?.();
      this.removeNotificationListener = undefined;
      rpcClient.close(new Error("Stagehand initialization failed", { cause: error }), {
        closeTransport: false,
      });
      this.rpcClient = undefined;
      throw error;
    }

    this.isInitialized = true;
    this.closePromise = undefined;
  }

  async act(instruction: string, options?: StagehandClientActOptions): Promise<ActResult>;
  async act(instruction: Action, options?: StagehandClientActOptions): Promise<ActResult>;
  async act(instruction: string | Action, options?: StagehandClientActOptions): Promise<ActResult> {
    const { page, ...clientOptions } = StagehandClientActOptionsSchema.parse(options ?? {});
    const targetPage = page ?? (await this.browser.context.activePage());
    if (!targetPage) throw new Error("Stagehand has no active page.");
    const protocolOptions = serializeClientLocatorOptions("act", targetPage.pageId, clientOptions);
    const response = await this.connectedRpcClient.send(StagehandMethods.stagehandAct, {
      pageId: targetPage.pageId,
      instruction,
      ...(options === undefined ? {} : { options: protocolOptions }),
    });

    return response;
  }

  async observe(
    instruction?: string,
    options?: StagehandClientObserveOptions,
  ): Promise<ObserveResult> {
    const { page, ...clientOptions } = StagehandClientObserveOptionsSchema.parse(options ?? {});
    const targetPage = page ?? (await this.browser.context.activePage());
    if (!targetPage) throw new Error("Stagehand has no active page.");
    const protocolOptions = serializeClientLocatorOptions(
      "observe",
      targetPage.pageId,
      clientOptions,
    );
    const response = await this.connectedRpcClient.send(StagehandMethods.stagehandObserve, {
      pageId: targetPage.pageId,
      ...(instruction === undefined ? {} : { instruction }),
      ...(options === undefined ? {} : { options: protocolOptions }),
    });

    return response;
  }

  async extract(
    instruction: string,
    options?: StagehandClientExtractOptions,
  ): Promise<ExtractResult<z.ZodType<DefaultExtractData>>>;
  async extract<Schema extends z.ZodType>(
    instruction: string,
    schema: Schema,
    options?: StagehandClientExtractOptions,
  ): Promise<ExtractResult<Schema>>;
  async extract<Schema extends z.ZodType | StagehandClientExtractOptions>(
    instruction: string,
    schema?: Schema,
    options?: StagehandClientExtractOptions,
  ): Promise<ExtractResult<z.ZodType>> {
    const hasCustomSchema = isZodSchema(schema);
    const resolvedSchema = hasCustomSchema ? schema : DefaultExtractDataSchema;
    const resolvedOptions = hasCustomSchema ? options : schema;
    const { page, ...clientOptions } = StagehandClientExtractOptionsSchema.parse(
      resolvedOptions ?? {},
    );
    const targetPage = page ?? (await this.browser.context.activePage());
    if (!targetPage) throw new Error("Stagehand has no active page.");
    const protocolOptions = serializeClientLocatorOptions(
      "extract",
      targetPage.pageId,
      clientOptions,
    );
    const response = await this.connectedRpcClient.send(StagehandMethods.stagehandExtract, {
      pageId: targetPage.pageId,
      instruction,
      ...(hasCustomSchema ? { schema: z.json().parse(z.toJSONSchema(resolvedSchema)) } : {}),
      ...(resolvedOptions === undefined ? {} : { options: protocolOptions }),
    });

    return {
      ...response,
      data: resolvedSchema.parse(response.data),
    };
  }

  close(): Promise<void> {
    this.closePromise ??= (async () => {
      let shouldReleaseBrowserClaim = !this.isInitialized;
      try {
        if (this.isInitialized) {
          try {
            await this.rpcClient?.send(StagehandMethods.stagehandClose, {});
            shouldReleaseBrowserClaim = true;
          } catch (error) {
            if (!(error instanceof CDPConnectionClosedError)) throw error;
            shouldReleaseBrowserClaim = true;
          }
        }
      } finally {
        this.removeClientLLMHandler?.();
        this.removeClientLLMHandler = undefined;
        this.removeNotificationListener?.();
        this.removeNotificationListener = undefined;
        this.rpcClient?.close(new Error("Stagehand closed"), { closeTransport: false });
        this.rpcClient = undefined;
        detachStagehandBrowserContext(this.browserHandle);
        if (shouldReleaseBrowserClaim) releaseStagehandBrowser(this.browserHandle);
        this.isInitialized = false;
      }
    })();
    return this.closePromise;
  }

  private get connectedRpcClient(): RPCClient {
    if (!this.isInitialized || !this.rpcClient) {
      throw new Error(
        "Stagehand is unavailable. Create a new instance with await Stagehand.create().",
      );
    }
    return this.rpcClient;
  }
}

function isDefinitiveRPCErrorResponse(error: unknown): boolean {
  return error instanceof Error && JSONRPCErrorObjectSchema.safeParse(error.cause).success;
}

function stagehandCreateParamsForWorker(
  createConfig: ResolvedStagehandClientCreateConfig,
  browser: ClaimedStagehandBrowser,
) {
  const { logging, model, ...protocolParams } = createConfig;
  const protocolModel = model && "generate" in model ? { source: "client" as const } : model;

  return StagehandInitParamsSchema.parse({
    protocolVersion: STAGEHAND_PROTOCOL_VERSION,
    clientInfo: STAGEHAND_SDK_CLIENT_INFO,
    browserCdpUrl: browser.cdpClient.webSocketDebuggerUrl,
    logLevel: logging.level,
    ...protocolParams,
    ...browser.workerInitMetadata,
    ...(protocolModel === undefined ? {} : { model: protocolModel }),
  });
}

const LOG_LEVEL_PRIORITY = {
  debug: 10,
  info: 20,
  warn: 30,
  error: 40,
  off: Number.POSITIVE_INFINITY,
} as const;

function handleStagehandNotification(
  notification: StagehandRpcNotification,
  logging: ResolvedStagehandClientLoggingConfig,
): void {
  if (notification.method !== StagehandNotifications.log.name) return;
  const log = notification.params;
  if (LOG_LEVEL_PRIORITY[log.level] < LOG_LEVEL_PRIORITY[logging.level]) return;

  process.stderr.write(renderStagehandLog(log, logging.format) + "\n");
  if (!logging.onLog) return;

  try {
    const result = logging.onLog(log);
    if (result && typeof result === "object" && "then" in result) {
      void Promise.resolve(result).catch(reportOnLogError);
    }
  } catch (error) {
    reportOnLogError(error);
  }
}

function renderStagehandLog(
  log: Extract<StagehandRpcNotification, { method: "stagehand.log" }>["params"],
  format: ResolvedStagehandClientLoggingConfig["format"],
): string {
  if (format === "json") return JSON.stringify(log);

  const data = Object.keys(log.data).length === 0 ? "" : ` ${JSON.stringify(log.data)}`;
  return `[stagehand] ${log.level.toUpperCase()} ${log.message}${data}`;
}

function reportOnLogError(error: unknown): void {
  const message = error instanceof Error ? error.message : String(error);
  process.stderr.write(`[stagehand] ERROR onLog callback failed: ${message}\n`);
}

File: packages/sdk-python/src/stagehand/stagehand.py
from __future__ import annotations

import asyncio
import builtins
import inspect
import json
import sys
from collections.abc import Callable, Mapping
from typing import TypeVar, cast, overload

from pydantic import BaseModel

from . import client_models as _client_models
from ._generated import models as _models
from ._generated.input_types import Action as ActionInput
from ._generated.input_types import Locator as ProtocolLocator
from ._generated.input_types import ModelConfig, TelemetryConfig, Variables
from ._generated.models import (
    Action,
    ActOptions,
    ActResult,
    CallbackBatchOptions,
    CallbackBatchParams,
    CallbackBatchResult,
    ClientModelReference,
    EmptyParams,
    ExtractOptions,
    FieldSchema0,
    LLMGenerateParams,
    LLMGenerateResult,
    ObserveOptions,
    ObserveResult,
    StagehandActParams,
    StagehandCloseResult,
    StagehandExtractParams,
    StagehandInitParams,
    StagehandInitResult,
    StagehandLog,
    StagehandMetrics,
    StagehandObserveParams,
)
from ._generated.protocol_version import STAGEHAND_PROTOCOL_VERSION
from ._sdk_identity import STAGEHAND_SDK_CLIENT_INFO
from .browser import (
    StagehandBrowser,
    _attach_browser_context,
    _claim_browser,
    _ClaimedBrowser,
    _detach_browser_context,
    _invalidate_browser,
    _release_browser,
)
from .browser_context import BrowserContext
from .cdp_client import CDPConnectionClosedError
from .client_models import (
    DefaultExtract,
    ExtractResult,
    _cache_config,
    _ExtractWireResult,
    _model_config,
)
from .client_types import Cache, LLMGenerateCallback, StagehandClientLoggingConfig
from .locator import Locator
from .page import Page
from .rpc_client import RPCClient, RPCError
from .timeouts import with_stagehand_init_deadline

ResultModel = TypeVar("ResultModel", bound=BaseModel)
_CONSTRUCTION_TOKEN = object()
_UNSET_BATCH_INPUT = object()
_MAX_CALLBACK_BATCH_TIMEOUT_MS = 2_147_483_647 - 10_000
_UNAVAILABLE_MESSAGE = (
    "Stagehand is unavailable. Create a new instance with await Stagehand.create()."
)


def _serialize_locator(locator: Locator, page_id: str, method: str) -> ProtocolLocator:
    if locator.page_id != page_id:
        raise TypeError(f"{method}() locator must belong to the target page")
    if locator.nth_index is not None:
        return ProtocolLocator(selector=locator.selector, nth=locator.nth_index)
    return ProtocolLocator(selector=locator.selector)


def _serialize_locators(
    locators: list[Locator] | None,
    page_id: str,
    method: str,
) -> list[ProtocolLocator] | None:
    if locators is None:
        return None
    return [_serialize_locator(locator, page_id, method) for locator in locators]


class Stagehand:
    def __init__(
        self,
        *,
        _token: object | None = None,
        browser: StagehandBrowser | None = None,
        create_config: _client_models.StagehandClientCreateConfig | None = None,
    ) -> None:
        if _token is not _CONSTRUCTION_TOKEN or browser is None or create_config is None:
            raise TypeError(
                "Stagehand cannot be constructed directly; use await Stagehand.create()"
            )
        self._browser_handle = browser
        self._create_config = create_config
        self._rpc_client: RPCClient | None = None
        self._remove_notification_listener: Callable[[], None] | None = None
        self._remove_client_llm_handler: Callable[[], None] | None = None
        self._initialized = False
        self._init_request_started = False
        self._close_task: asyncio.Task[None] | None = None

    @classmethod
    async def create(
        cls,
        *,
        browser: StagehandBrowser,
        api_key: str | None = None,
        api_url: str | None = None,
        model: str | LLMGenerateCallback | None = None,
        model_api_key: str | None = None,
        model_headers: Mapping[str, str] | None = None,
        telemetry: TelemetryConfig | None = None,
        system_prompt: str | None = None,
        self_heal: bool | None = None,
        dom_settle_timeout_ms: int | None = None,
        cache: Cache | None = None,
        logging: StagehandClientLoggingConfig | None = None,
    ) -> Stagehand:
        if not isinstance(browser, StagehandBrowser):
            raise TypeError("browser must be created by local_browser or browserbase")
        model_connection_options = (model_api_key, model_headers)
        if model is None and any(value is not None for value in model_connection_options):
            raise TypeError("model connection options require a model name")
        if callable(model) and any(value is not None for value in model_connection_options):
            raise TypeError("model connection options cannot be used with an LLM callback")

        resolved_model: _models.ModelConfig | _client_models.ClientLLM | None
        if isinstance(model, str):
            resolved_model = _model_config(
                model,
                api_key=model_api_key,
                headers=dict(model_headers) if model_headers is not None else None,
            )
        elif model is not None:
            resolved_model = _client_models.ClientLLM(generate=model)
        else:
            resolved_model = None

        values: dict[str, object] = {
            name: value
            for name, value in (
                ("api_key", api_key),
                ("api_url", api_url),
                ("system_prompt", system_prompt),
                ("self_heal", self_heal),
                ("dom_settle_timeout_ms", dom_settle_timeout_ms),
                ("cache", _cache_config(cache) if cache is not None else None),
                ("logging", logging),
            )
            if value is not None
        }
        if resolved_model is not None:
            values["model"] = resolved_model
        if telemetry is not None:
            values["telemetry"] = telemetry
        create_config = _client_models.StagehandClientCreateConfig.model_validate(values)
        claimed = _claim_browser(browser)
        stagehand = cls(
            _token=_CONSTRUCTION_TOKEN,
            browser=browser,
            create_config=create_config,
        )
        try:
            await with_stagehand_init_deadline(stagehand._initialize(claimed))
        except BaseException as error:
            fail_closed = stagehand._initialization_failure_is_ambiguous(error)
            await asyncio.shield(stagehand._cleanup_failed_create(browser, fail_closed=fail_closed))
            raise
        return stagehand

    @property
    def browser(self) -> StagehandBrowser:
        return self._browser_handle

    @property
    def initialized(self) -> bool:
        return self._initialized

    async def metrics(self) -> StagehandMetrics:
        return await self._connected_rpc_client.send(
            "stagehand.metrics",
            EmptyParams(),
            StagehandMetrics,
        )

    async def experimental_batch(
        self,
        source: str,
        input: object = _UNSET_BATCH_INPUT,
        *,
        timeout: int = 30_000,
        page: Page | None = None,
    ) -> object:
        """Run trusted JavaScript against the worker-local Stagehand object model."""
        if not isinstance(source, str) or not source.strip():
            raise TypeError("source must be a non-empty JavaScript string")
        if isinstance(timeout, bool) or not isinstance(timeout, int) or timeout <= 0:
            raise ValueError("timeout must be a positive number of milliseconds")
        if timeout > _MAX_CALLBACK_BATCH_TIMEOUT_MS:
            raise ValueError(
                f"timeout must not exceed {_MAX_CALLBACK_BATCH_TIMEOUT_MS} milliseconds"
            )
        input_model: _models.FieldSchema2 | None = None
        if input is not _UNSET_BATCH_INPUT:
            try:
                serialized_input = json.dumps(input, allow_nan=False)
                input_model = _models.FieldSchema2.model_validate_json(
                    serialized_input,
                    strict=True,
                )
            except (TypeError, ValueError) as error:
                raise TypeError("input must be JSON-serializable") from error
        options_model = CallbackBatchOptions(
            **({"page_id": page.page_id} if page is not None else {}),
            timeout=timeout,
        )
        params = (
            CallbackBatchParams(
                callback_source=source,
                options=options_model,
            )
            if input is _UNSET_BATCH_INPUT
            else CallbackBatchParams(
                callback_source=source,
                input=input_model,
                options=options_model,
            )
        )
        result = await self._connected_rpc_client.send(
            "stagehand.callback_batch",
            params,
            CallbackBatchResult,
        )
        return (
            result.value.model_dump(mode="json")
            if isinstance(result.value, BaseModel)
            else result.value
        )

    async def _initialize(self, claimed: _ClaimedBrowser) -> None:
        rpc_client = RPCClient(claimed.cdp_client)
        self._rpc_client = rpc_client
        self._remove_notification_listener = rpc_client.on_notification(
            "stagehand.log",
            StagehandLog,
            self._handle_stagehand_notification,
        )
        client_llm = self._create_config.model
        if isinstance(client_llm, _client_models.ClientLLM):

            async def generate(params: LLMGenerateParams) -> LLMGenerateResult:
                return LLMGenerateResult(root=await client_llm.generate(params.root))

            self._remove_client_llm_handler = rpc_client.on_request(
                "llm.generate",
                LLMGenerateParams,
                LLMGenerateResult,
                generate,
            )

        init_params = self._worker_init_params(claimed)
        self._init_request_started = True
        await rpc_client.send(
            "stagehand.init",
            init_params,
            StagehandInitResult,
        )
        _attach_browser_context(
            self._browser_handle,
            BrowserContext(rpc_client, self._browser_handle.close),
        )
        self._initialized = True

    async def act(
        self,
        instruction: str | ActionInput | Action,
        *,
        page: Page | None = None,
        model: ModelConfig | None = None,
        variables: Variables | None = None,
        timeout: float | None = None,
        locator: Locator | None = None,
        ignore_locators: list[Locator] | None = None,
        cache: Cache | None = None,
    ) -> ActResult:
        target_page = page or await self.browser.context.active_page()
        if target_page is None:
            raise RuntimeError("Stagehand has no active page")
        options = ActOptions.model_validate({
            name: value
            for name, value in (
                ("model", model),
                ("variables", variables),
                ("timeout", timeout),
                (
                    "locator",
                    _serialize_locator(locator, target_page.page_id, "act")
                    if locator is not None
                    else None,
                ),
                (
                    "ignore_locators",
                    _serialize_locators(ignore_locators, target_page.page_id, "act"),
                ),
                ("cache", _cache_config(cache) if cache is not None else None),
            )
            if value is not None
        })
        params = StagehandActParams.model_validate({
            "page_id": target_page.page_id,
            "instruction": instruction,
        })
        if options.model_fields_set:
            params.options = options
        result = await self._connected_rpc_client.send("stagehand.act", params, ActResult)
        return result

    async def observe(
        self,
        instruction: str | None = None,
        *,
        page: Page | None = None,
        model: ModelConfig | None = None,
        variables: Variables | None = None,
        timeout: float | None = None,
        locator: Locator | None = None,
        ignore_locators: list[Locator] | None = None,
        cache: Cache | None = None,
    ) -> ObserveResult:
        target_page = page or await self.browser.context.active_page()
        if target_page is None:
            raise RuntimeError("Stagehand has no active page")
        options = ObserveOptions.model_validate({
            name: value
            for name, value in (
                ("model", model),
                ("variables", variables),
                ("timeout", timeout),
                (
                    "locator",
                    _serialize_locator(locator, target_page.page_id, "observe")
                    if locator is not None
                    else None,
                ),
                (
                    "ignore_locators",
                    _serialize_locators(ignore_locators, target_page.page_id, "observe"),
                ),
                ("cache", _cache_config(cache) if cache is not None else None),
            )
            if value is not None
        })
        params = StagehandObserveParams(page_id=target_page.page_id, instruction=instruction)
        if options.model_fields_set:
            params.options = options
        result = await self._connected_rpc_client.send("stagehand.observe", params, ObserveResult)
        return result

    @overload
    async def extract(
        self,
        instruction: str,
        schema: builtins.type[DefaultExtract] = DefaultExtract,
        *,
        page: Page | None = None,
        model: ModelConfig | None = None,
        timeout: float | None = None,
        screenshot: bool | None = None,
        locator: Locator | None = None,
        ignore_locators: list[Locator] | None = None,
        cache: Cache | None = None,
    ) -> ExtractResult[DefaultExtract]: ...

    @overload
    async def extract(
        self,
        instruction: str,
        schema: builtins.type[ResultModel],
        *,
        page: Page | None = None,
        model: ModelConfig | None = None,
        timeout: float | None = None,
        screenshot: bool | None = None,
        locator: Locator | None = None,
        ignore_locators: list[Locator] | None = None,
        cache: Cache | None = None,
    ) -> ExtractResult[ResultModel]: ...

    async def extract(
        self,
        instruction: str,
        schema: builtins.type[ResultModel] = cast(builtins.type[ResultModel], DefaultExtract),
        *,
        page: Page | None = None,
        model: ModelConfig | None = None,
        timeout: float | None = None,
        screenshot: bool | None = None,
        locator: Locator | None = None,
        ignore_locators: list[Locator] | None = None,
        cache: Cache | None = None,
    ) -> ExtractResult[ResultModel]:
        target_page = page or await self.browser.context.active_page()
        if target_page is None:
            raise RuntimeError("Stagehand has no active page")
        options = ExtractOptions.model_validate({
            name: value
            for name, value in (
                ("model", model),
                ("timeout", timeout),
                ("screenshot", screenshot),
                (
                    "locator",
                    _serialize_locator(locator, target_page.page_id, "extract")
                    if locator is not None
                    else None,
                ),
                (
                    "ignore_locators",
                    _serialize_locators(ignore_locators, target_page.page_id, "extract"),
                ),
                ("cache", _cache_config(cache) if cache is not None else None),
            )
            if value is not None
        })
        params = StagehandExtractParams(
            page_id=target_page.page_id,
            instruction=instruction,
        )
        if schema is not DefaultExtract:
            params.schema_ = FieldSchema0.model_validate(schema.model_json_schema())
        if options.model_fields_set:
            params.options = options
        result = await self._connected_rpc_client.send(
            "stagehand.extract", params, _ExtractWireResult
        )
        return ExtractResult(
            data=schema.model_validate(result.data),
            metadata=result.metadata,
        )

    async def close(self) -> None:
        async def close_impl() -> None:
            release_browser_claim = not self._initialized
            try:
                if self._initialized and self._rpc_client is not None:
                    try:
                        await self._rpc_client.send(
                            "stagehand.close",
                            EmptyParams(),
                            StagehandCloseResult,
                        )
                        release_browser_claim = True
                    except CDPConnectionClosedError:
                        release_browser_claim = True
            finally:
                try:
                    await asyncio.shield(self._release_resources())
                finally:
                    if release_browser_claim:
                        _release_browser(self._browser_handle)

        if self._close_task is None:
            self._close_task = asyncio.create_task(
                close_impl(),
                name="stagehand-close",
            )
        await asyncio.shield(self._close_task)

    @property
    def _connected_rpc_client(self) -> RPCClient:
        if not self._initialized or self._rpc_client is None:
            raise RuntimeError(_UNAVAILABLE_MESSAGE)
        return self._rpc_client

    def _worker_init_params(self, claimed: _ClaimedBrowser) -> StagehandInitParams:
        browser_cdp_url = claimed.cdp_client.web_socket_debugger_url
        if browser_cdp_url is None:
            raise RuntimeError("The browser CDP WebSocket URL is unavailable")
        values = self._create_config.model_dump(
            exclude={"logging", "model"},
            exclude_unset=True,
        )
        if isinstance(self._create_config.model, _client_models.ClientLLM):
            values["model"] = ClientModelReference(source="client")
        elif self._create_config.model is not None:
            values["model"] = self._create_config.model
        values["protocol_version"] = STAGEHAND_PROTOCOL_VERSION
        values["client_info"] = STAGEHAND_SDK_CLIENT_INFO
        values["browser_cdp_url"] = browser_cdp_url
        values["log_level"] = self._create_config.logging.level
        metadata = claimed.worker_init_metadata
        if metadata.api_key is not None:
            values["api_key"] = metadata.api_key
        if metadata.browser is not None:
            values["browser"] = metadata.browser
        return StagehandInitParams.model_validate(values)

    async def _handle_stagehand_notification(self, notification: StagehandLog) -> None:
        logging = self._create_config.logging
        if not _is_log_level_enabled(notification.level.value, logging.level):
            return

        sys.stderr.write(_render_stagehand_log(notification, logging.format) + "\n")
        if logging.on_log is None:
            return

        try:
            result = logging.on_log(notification)
            if inspect.isawaitable(result):
                await result
        except Exception as error:
            sys.stderr.write(f"[stagehand] ERROR on_log callback failed: {error}\n")

    async def _release_resources(self) -> None:
        if self._remove_client_llm_handler is not None:
            self._remove_client_llm_handler()
            self._remove_client_llm_handler = None
        if self._remove_notification_listener is not None:
            self._remove_notification_listener()
            self._remove_notification_listener = None
        rpc_client = self._rpc_client
        self._rpc_client = None
        _detach_browser_context(self._browser_handle)
        self._initialized = False
        if rpc_client is not None:
            await rpc_client.close(RuntimeError("Stagehand closed"), close_transport=False)

    def _initialization_failure_is_ambiguous(self, error: BaseException) -> bool:
        if isinstance(error, (asyncio.CancelledError, TimeoutError)):
            return True
        return self._init_request_started and not isinstance(error, RPCError)

    async def _cleanup_failed_create(
        self,
        browser: StagehandBrowser,
        *,
        fail_closed: bool,
    ) -> None:
        if fail_closed:
            browser_invalidation = _invalidate_browser(browser)
            try:
                await self._release_resources()
            finally:
                await browser_invalidation
            return

        try:
            await self._release_resources()
        finally:
            _release_browser(browser)


_LOG_LEVEL_PRIORITY = {
    "debug": 10,
    "info": 20,
    "warn": 30,
    "error": 40,
    "off": float("inf"),
}


def _is_log_level_enabled(level: str, threshold: str) -> bool:
    return _LOG_LEVEL_PRIORITY[level] >= _LOG_LEVEL_PRIORITY[threshold]


def _render_stagehand_log(notification: StagehandLog, format_: str) -> str:
    data = notification.data.model_dump(mode="json")
    record = {
        "level": notification.level.value,
        "message": notification.message,
        "data": data,
    }
    if format_ == "json":
        return json.dumps(record, separators=(",", ":"))

    suffix = "" if not data else f" {json.dumps(data, separators=(',', ':'))}"
    return f"[stagehand] {notification.level.value.upper()} {notification.message}{suffix}"
